composition-js/src/validate.ts

```diff
@@ -474,7 +491,20 @@ export class ValidationState {
       return { error: satisfiabilityError(newPath, this.subgraphPaths.map((p) => p.path), deadEnds) };
     }
 
-    const updatedState = new ValidationState(newPath, newSubgraphPaths);
+    // If the edge has an override condition, we should capture it in the state so
+    // that we can ignore later edges that don't satisfy the condition.
+    const newOverrideConditions = new Map([...this.selectedOverrideConditions]);
+    if (supergraphEdge.overrideCondition) {
+      newOverrideConditions.set(
+        supergraphEdge.overrideCondition.label,
+        supergraphEdge.overrideCondition.condition
+      );
+    }
+    const updatedState = new ValidationState(
+      newPath,
+      newSubgraphPaths,
+      newOverrideConditions,
+    );
```

Comment:
These `newOverrideConditions` need to be computed prior to the call to `advancePathWithTransition()`, and they should be passed into `advancePathWithTransition()` instead of `this.selectedOverrideConditions`.

Reply:
[2391f43](https://github.com/apollographql/federation/pull/2902/commits/2391f43bfcb9e698f38ee7da77dee9696076c82c)